File: android/src/main/java/com/margelo/nitro/xprinter/HybridPrinterImageFactory.kt
package com.margelo.nitro.xprinter

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlin.math.floor

/**
 * The Android implementation of [PrinterImageFactory].
 *
 * Decoding and dithering are CPU-bound and can run for hundreds of milliseconds
 * on a full-width image, so they own an IO scope and never touch the main thread.
 */
@DoNotStrip
class HybridPrinterImageFactory : HybridPrinterImageFactorySpec() {
  private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

  override fun rasterize(options: RasterizeOptions): Promise<HybridPrinterRasterSpec> =
    Promise.async(ioScope) {
      require(options.widthDots >= 1 && options.widthDots == floor(options.widthDots)) {
        "widthDots must be a positive whole number of dots (got ${options.widthDots})."
      }
      val context = requireContext()

      val widthDots = options.widthDots.toInt()
      val bitmap = if (SourceFiles.isPdf(context, options.source)) {
        // Rasterize straight to the final fitted size — see PdfPageRenderer.
        PdfPageRenderer.render(
          context,
          options.source,
          options.pageIndex?.toInt() ?: 0,
          widthDots,
          options.maxHeightDots?.toInt()
        )
      } else {
        SourceImageLoader.load(context, options.source)
      }

      try {
        HybridPrinterRaster(MonochromeRasterizer.rasterize(bitmap, options))
      } finally {
        bitmap.recycle()
      }
    }

  override fun countPages(source: String): Promise<Double> =
    Promise.async(ioScope) {
      val context = requireContext()
      if (SourceFiles.isPdf(context, source)) {
        PdfPageRenderer.pageCount(context, source).toDouble()
      } else {
        1.0
      }
    }

  private fun requireContext() = NitroModules.applicationContext
    ?: throw IllegalStateException(
      "No ReactApplicationContext is available yet. Use PrinterImages after " +
        "React Native has finished initializing."
    )
}

File: ios/HybridPrinterImageFactory.swift
import CoreGraphics
import Foundation
import NitroModules

/// The iOS implementation of `PrinterImageFactory`.
///
/// Decoding and dithering are CPU-bound and can run for hundreds of
/// milliseconds on a full-width image, so they run on their own queue and
/// never touch the main thread.
final class HybridPrinterImageFactory: HybridPrinterImageFactorySpec {
  private let queue = DispatchQueue(label: "com.margelo.nitro.xprinter.image-factory", qos: .userInitiated)

  func rasterize(options: RasterizeOptions) throws -> Promise<(any HybridPrinterRasterSpec)> {
    Promise.parallel(queue) {
      guard options.widthDots >= 1, options.widthDots == options.widthDots.rounded(.down) else {
        throw RuntimeError.error(
          withMessage: "widthDots must be a positive whole number of dots (got \(options.widthDots))."
        )
      }

      // Rasterize a PDF page straight to the final fitted size — see
      // PdfPageRenderer — rather than decoding it like a bitmap image.
      let image: CGImage =
        if try SourceFiles.isPdf(source: options.source) {
          try PdfPageRenderer.render(
            source: options.source,
            pageIndex: options.pageIndex.map { Int($0) } ?? 0,
            targetWidthDots: Int(options.widthDots),
            maxHeightDots: options.maxHeightDots.map { Int($0) },
            trimToContent: options.trimToContent
          )
        } else {
          try SourceImageLoader.load(source: options.source)
        }

      let raster = try MonochromeRasterizer.rasterize(image: image, options: options)
      return HybridPrinterRaster(raster: raster) as any HybridPrinterRasterSpec
    }
  }

  func countPages(source: String) throws -> Promise<Double> {
    Promise.parallel(queue) {
      if try SourceFiles.isPdf(source: source) {
        return Double(try PdfPageRenderer.pageCount(source: source))
      }
      return 1.0
    }
  }
}
